skillmap/src/lib/codeCarryover.ts

```diff
@@ -0,0 +1,185 @@
+import { lookupPreviousActivityStates } from "./skillMapUtils";
+import { getProjectAsync, saveProjectAsync } from "./workspaceProvider";
+
+export async function carryoverProjectCode(user: UserState, pageSource: string, map: SkillMap, activityId: string, carryoverCode: boolean) {
+    const progress = user.mapProgress[pageSource][map.mapId];
+
+    const headerId = progress.activityState[activityId]?.headerId;
+
+    const previous = lookupPreviousActivityStates(user, pageSource, map, activityId);
+    const previousHeaderId = previous.find(state => state?.isCompleted &&
```

Comment:
To confirm my reading of this; `lookupPreviousActivityStates` returns an array of elements that could potentially unlock the current activity (so parent nodes in the graph, not grandparents+), and this will grab the first one from that list that is completed.

This definitely falls into future work as the rendering is not supported yet, but out of curiosity -- in the future when shannon's stuff comes in and we can have skillmaps that have 2 parents to an activity, if both are completed then I guess the expectation would be that this would pull from first one defined in the markdown, and maybe allow an override as a property in the markdown (by specifying the id of the activity to pull from just in general I suppose) -- sound about right? Something to discuss in future I suppose.

Reply:
in my ideal world we'd have UI to let the user choose, or we would take the more recently opened activity (along with displaying the activity name--"You are importing code from [Creating Enemies]" eg). but yeah definitely worth stepping through the flow as a team once we have all the pieces together

Reply:
yeah this is absolutely just a placeholder for now. my plan was to do it based on whichever was most recent completed, but ui would be cool also.